File: problems/brace_expansion/brace_expansion.py
import re

from typing import List


def braceExpansion(s: str) -> List[str]:
    results: List[str] = []

    def backtrack(prefixes: List[str], remaining: str) -> None:
        """
        Recursive helper that expands the string.

        prefixes : partial expansions built so far
        remaining: the remaining part of the string to process
        """
        nonlocal results

        # Regex captures:
        # 1) leading literal before {   ([^{}]*)
        # 2) choices inside {}          ([^}]*)
        # 3) literal after }            ([^{}]*)
        # 4) remaining tail             (.*)
        match = re.search(r"([^{}]*)\{([^}]*)\}([^{}]*)(.*)", remaining)

        if not match:
            # Base case: no more braces left
            if prefixes:
                # Append remaining to each existing prefix
                results.extend(prefix + remaining for prefix in prefixes)
            else:
                # No prefix: just add the raw remaining string
                results.append(remaining)
            return

        lead, choices_str, mid, tail = match.groups()
        choices = sorted(choices_str.split(","))  # expand into sorted list of choices

        # If no prefixes yet, start from empty string
        base_prefixes = prefixes if prefixes else [""]
        new_prefixes: List[str] = []

        # Build new partial expansions by combining each prefix with each choice
        for prefix in base_prefixes:
            for choice in choices:
                new_prefixes.append(prefix + lead + choice + mid)

        # Recurse on the tail of the string
        backtrack(new_prefixes, tail)

    # Start recursion with empty prefix and full string
    backtrack([], s)

    return results

File: problems/brace_expansion/brace_expansion.cc
#include "brace_expansion.h"

#include <algorithm>
#include <cstddef>
#include <regex>
#include <string>
#include <utility>
#include <vector>

namespace {

std::vector<std::string> splitByComma(const std::string& str) {
    std::vector<std::string> out;
    std::size_t start = 0;
    while (true) {
        const std::size_t pos = str.find(',', start);
        if (pos == std::string::npos) {
            out.emplace_back(str.substr(start));
            break;
        }
        out.emplace_back(str.substr(start, pos - start));
        start = pos + 1;
    }
    return out;
}

}  // namespace

std::vector<std::string> braceExpansion(const std::string& input) {
    std::vector<std::string> prefixes;
    prefixes.emplace_back("");  // Start with an empty prefix
    std::string remaining = input;

    // Regex captures (compile once; reuse):
    // 1) leading literal before {   ([^{}]*)
    // 2) choices inside {}          ([^}]*)
    // 3) literal after }            ([^{}]*)
    // 4) remaining tail             (.*)
    static const std::regex REGEX(R"(([^{}]*)\{([^}]*)\}([^{}]*)(.*))");

    std::smatch match;
    // Iterate group by group until no more braces
    while (std::regex_match(remaining, match, REGEX)) {
        const std::string& lead = match[1].str();
        const std::string& choicesStr = match[2].str();
        const std::string& mid = match[3].str();
        std::string tail = match[4].str();  // will become new remaining

        // Expand choices and sort them
        std::vector<std::string> choices = splitByComma(choicesStr);
        if (choices.size() > 1) {
            std::ranges::sort(choices);
        }

        std::vector<std::string> newPrefixes;
        newPrefixes.reserve(prefixes.size() * std::max(choices.size(), 1UL));
        for (const auto& prefix : prefixes) {
            for (const auto& choice : choices) {
                std::string newPrefix = prefix;
                newPrefix += lead;
                newPrefix += choice;
                newPrefix += mid;
                newPrefixes.emplace_back(std::move(newPrefix));
            }
        }

        prefixes.swap(newPrefixes);  // reuse memory
        remaining.swap(tail);        // continue with the tail
    }

    // Append final remaining tail (no more braces)
    std::vector<std::string> results;
    results.reserve(prefixes.size());
    if (!prefixes.empty()) {
        for (const auto& prefix : prefixes) {
            std::string result = prefix;
            result += remaining;
            results.emplace_back(std::move(result));
        }
    } else {
        results.emplace_back(remaining);  // input had no braces
    }

    return results;
}
